SelectWithSelectorFix_TcpIpConnection_TransferStressTest.testTinyPackets

master (commit a32c26cc24130659320a1aa8aa4940b53062ec35)

http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/288/testReport/com.hazelcast.internal.networking.nio/SelectWithSelectorFix_TcpIpConnection_TransferStressTest/testTinyPackets/

**Stacktrace**
```

java.lang.AssertionError: expected:<155083074> but was:<155074074>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.internal.nio.tcp.TcpIpConnection_AbstractTransferStressTest$1.run(TcpIpConnection_AbstractTransferStressTest.java:130)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1325)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1342)
	at com.hazelcast.internal.nio.tcp.TcpIpConnection_AbstractTransferStressTest.testPackets(TcpIpConnection_AbstractTransferStressTest.java:120)
	at com.hazelcast.internal.nio.tcp.TcpIpConnection_AbstractTransferStressTest.testPackets(TcpIpConnection_AbstractTransferStressTest.java:91)
	at com.hazelcast.internal.nio.tcp.TcpIpConnection_AbstractTransferStressTest.testTinyPackets(TcpIpConnection_AbstractTransferStressTest.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/hazelcast/src/main/java/com/hazelcast/client/properties/ClientProperty.java
+++ b/hazelcast/src/main/java/com/hazelcast/client/properties/ClientProperty.java
@@ -164,12 +164,12 @@ public final class ClientProperty {
      * can be done correctly. This property sets the window the concurrency detection will signalling
      * that concurrency has been detected, even if there are no further updates in that window.
      *
-     * Normally in a concurrency system the windows keeps sliding forward so it will always remain
+     * Normally in a concurrent system the window keeps sliding forward so it will always remain
      * concurrent.
      *
-     * Setting it too high effectively disabled the optimization because once concurrency has been detected
+     * Setting it too high effectively disables the optimization because once concurrency has been detected
      * it will keep that way. Setting it too low could lead to suboptimal performance because the system
-     * will try write through and other optimization even though the system is concurrent.
+     * will try write through and other optimizations even though the system is concurrent.
      */
     public static final HazelcastProperty CONCURRENT_WINDOW_MS
             = new HazelcastProperty("hazelcast.client.concurrent.window.ms", 100, MILLISECONDS);

--- a/hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioNetworking.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioNetworking.java
@@ -50,6 +50,7 @@ import java.util.logging.Level;
 
 import static com.hazelcast.internal.networking.nio.SelectorMode.SELECT;
 import static com.hazelcast.internal.networking.nio.SelectorMode.SELECT_NOW_STRING;
+import static com.hazelcast.internal.networking.nio.SelectorMode.SELECT_WITH_FIX;
 import static com.hazelcast.internal.nio.IOUtil.closeResource;
 import static com.hazelcast.internal.util.HashUtil.hashToIndex;
 import static com.hazelcast.internal.util.ThreadUtil.createThreadPoolName;
@@ -136,8 +137,20 @@ public final class NioNetworking implements Networking, DynamicMetricsProvider {
         this.selectorWorkaroundTest = ctx.selectorWorkaroundTest;
         this.idleStrategy = ctx.idleStrategy;
         this.concurrencyDetection = ctx.concurrencyDetection;
-        this.writeThroughEnabled = ctx.writeThroughEnabled;
-        this.selectionKeyWakeupEnabled = ctx.selectionKeyWakeupEnabled;
+        // selector mode SELECT_WITH_FIX requires that a single thread
+        // accesses a selector & its selectionKeys. Selection key wake-up
+        // and write through break this requirement, therefore must be
+        // disabled with SELECT_WITH_FIX.
+        this.writeThroughEnabled = ctx.writeThroughEnabled && selectorMode != SELECT_WITH_FIX;
+        this.selectionKeyWakeupEnabled = ctx.selectionKeyWakeupEnabled && selectorMode != SELECT_WITH_FIX;
+        if (selectorMode == SELECT_WITH_FIX
+                && (ctx.writeThroughEnabled || ctx.selectionKeyWakeupEnabled)) {
+            logger.warning("Selector mode SELECT_WITH_FIX is incompatible with write-through and selection key wakeup "
+                    + "optimizations and they have been disabled. Start Hazelcast with options "
+                    + "\"-Dhazelcast.io.selectionKeyWakeupEnabled=false -Dhazelcast.io.write.through=false\" to "
+                    + "explicitly disable selection key wakeup and write-through optimizations and avoid logging this "
+                    + "warning.");
+        }
     }
 
     @SuppressFBWarnings(value = "EI_EXPOSE_REP", justification = "used only for testing")
@@ -373,6 +386,15 @@ public final class NioNetworking implements Networking, DynamicMetricsProvider {
         context.collect(descriptorTcp, this);
     }
 
+    // package private accessors for testing
+    boolean isWriteThroughEnabled() {
+        return writeThroughEnabled;
+    }
+
+    boolean isSelectionKeyWakeupEnabled() {
+        return selectionKeyWakeupEnabled;
+    }
+
     private class ChannelCloseListenerImpl implements ChannelCloseListener {
         @Override
         public void onClose(Channel channel) {
@@ -465,8 +487,9 @@ public final class NioNetworking implements Networking, DynamicMetricsProvider {
             }
         }
 
-        public void setSelectionKeyWakeupEnabled(boolean selectionKeyWakeupEnabled) {
+        public Context selectionKeyWakeupEnabled(boolean selectionKeyWakeupEnabled) {
             this.selectionKeyWakeupEnabled = selectionKeyWakeupEnabled;
+            return this;
         }
 
         public Context writeThroughEnabled(boolean writeThroughEnabled) {

--- a/hazelcast/src/main/java/com/hazelcast/spi/properties/ClusterProperty.java
+++ b/hazelcast/src/main/java/com/hazelcast/spi/properties/ClusterProperty.java
@@ -292,12 +292,12 @@ public final class ClusterProperty {
      * This property sets the window the concurrency detection will signalling
      * that concurrency has been detected, even if there are no further updates in that window.
      *
-     * Normally in a concurrency system the windows keeps sliding forward so it will always remain
+     * Normally in a concurrent system the window keeps sliding forward so it will always remain
      * concurrent.
      *
-     * Setting it too high effectively disabled the optimization because once concurrency has been detected
+     * Setting it too high effectively disables the optimization because once concurrency has been detected
      * it will keep that way. Setting it too low could lead to suboptimal performance because the system
-     * will try write through and other optimization even though the system is concurrent.
+     * will try write through and other optimizations even though the system is concurrent.
      */
     public static final HazelcastProperty CONCURRENT_WINDOW_MS
             = new HazelcastProperty("hazelcast.concurrent.window.ms", 100, MILLISECONDS);

--- a/hazelcast/src/test/java/com/hazelcast/internal/networking/nio/NioNetworkingConfigurationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/networking/nio/NioNetworkingConfigurationTest.java
@@ -1,0 +1,94 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.networking.nio;
+
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.LoggingService;
+import com.hazelcast.test.HazelcastParallelParametersRunnerFactory;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.mockito.ArgumentMatchers;
+
+import java.util.Arrays;
+import java.util.Collection;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+// Tests configuration of write-through and selection key wake-up optimizations
+// with different selector modes.
+@RunWith(Parameterized.class)
+@Parameterized.UseParametersRunnerFactory(HazelcastParallelParametersRunnerFactory.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class NioNetworkingConfigurationTest {
+
+    @Parameterized.Parameters(name = "selectorMode={0},configured={1},expected={2}")
+    public static Collection<Object[]> parameters() {
+        return Arrays.asList(new Object[][]
+                {
+                        // selector mode, configured value, expected value
+                        {SelectorMode.SELECT, false, false},
+                        {SelectorMode.SELECT, true, true},
+                        {SelectorMode.SELECT_NOW, false, false},
+                        {SelectorMode.SELECT_NOW, true, true},
+                        {SelectorMode.SELECT_WITH_FIX, false, false},
+                        {SelectorMode.SELECT_WITH_FIX, true, false},
+                }
+        );
+    }
+
+    @Parameterized.Parameter
+    public SelectorMode selectorMode;
+
+    @Parameterized.Parameter(1)
+    public boolean configuredValue;
+
+    @Parameterized.Parameter(2)
+    public boolean expectedValue;
+
+    private NioNetworking.Context ctx;
+    private LoggingService loggingService;
+    private ILogger logger;
+
+    @Before
+    public void setup() {
+        loggingService = mock(LoggingService.class);
+        logger = mock(ILogger.class);
+        when(loggingService.getLogger(ArgumentMatchers.any(Class.class)))
+                .thenReturn(logger);
+
+        ctx = new NioNetworking.Context();
+        ctx.loggingService(loggingService);
+    }
+
+    @Test
+    public void testOptimizationsConfiguration() {
+        ctx.writeThroughEnabled(configuredValue)
+           .selectionKeyWakeupEnabled(configuredValue)
+           .selectorMode(selectorMode);
+        NioNetworking networking = new NioNetworking(ctx);
+
+        Assert.assertEquals(expectedValue, networking.isSelectionKeyWakeupEnabled());
+        Assert.assertEquals(expectedValue, networking.isWriteThroughEnabled());
+    }
+}

--- a/hazelcast/src/test/java/com/hazelcast/internal/networking/nio/SelectWithSelectorFix_TcpIpConnection_BasicTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/networking/nio/SelectWithSelectorFix_TcpIpConnection_BasicTest.java
@@ -20,13 +20,11 @@ import com.hazelcast.internal.nio.tcp.TcpIpConnection_AbstractBasicTest;
 import com.hazelcast.test.HazelcastSerialClassRunner;
 import com.hazelcast.test.annotation.QuickTest;
 import org.junit.Before;
-import org.junit.Ignore;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
 @RunWith(HazelcastSerialClassRunner.class)
 @Category(QuickTest.class)
-@Ignore
 public class SelectWithSelectorFix_TcpIpConnection_BasicTest
         extends TcpIpConnection_AbstractBasicTest {
 

